src/node/internal/internal_http_server.ts

```diff
@@ -0,0 +1,735 @@
+      options = {};
+    } else if (options == null) {
+      options = {};
+    } else {
+      validateObject(options, 'options');
+    }
+
+    if (requestListener) {
+      this.on('request', requestListener);
+    }
+
+    this.on('listening', setupConnectionsTracking);
+
+    this[kUniqueHeaders] = parseUniqueHeadersOption(
+      options.uniqueHeaders as (string | string[])[]
+    );
+  }
+
+  close(): this {
+    httpServerPreClose(this);
+    if (this.port) {
+      portMapper.delete(this.port);
+    }
+    return this;
+  }
+
+  closeAllConnections(): void {
+    // It doesn't make sense to support this method.
+    // Leave it as a noop.
+  }
+
+  closeIdleConnections(): void {
+    // It doesn't make sense to support this method.
+    // Leave it as a noop.
+  }
+
+  setTimeout(
+    msecs?: number | ((socket: Socket) => void),
+    callback?: (socket: Socket) => void
+  ): this {
+    if (typeof msecs === 'function') {
+      callback = msecs;
+      msecs = undefined;
+    } else if (typeof msecs === 'number') {
+      this.timeout = msecs;
+    }
+
+    if (typeof callback === 'function') {
+      this.once('timeout', callback);
+    }
+    return this;
+  }
+
+  async #onRequest(
+    request: Request,
+    _env: unknown,
+    _ctx: unknown
+  ): Promise<Response> {
+    const { incoming, response } = this.#toReqRes(request);
+    this.emit('connection', this, incoming);
+    this.emit('request', incoming, response);
+    return getServerResponseFetchResponse(response);
+  }
+
+  #toReqRes(request: Request): {
+    incoming: IncomingMessage;
+    response: ServerResponse;
+  } {
+    const incoming = new IncomingMessage();
+    const reqUrl = new URL(request.url);
+    incoming.url = reqUrl.pathname + reqUrl.search;
+
+    const headers = [];
+    for (const [key, value] of request.headers) {
+      headers.push(key, value);
+    }
+    incoming._addHeaderLines(headers, headers.length);
+
+    incoming.method = request.method;
+    incoming._stream = request.body;
+
+    const response = new ServerResponse(incoming);
+    return { incoming, response };
+  }
+
+  listen(...args: unknown[]): this {
+    const [options, callback] = _normalizeArgs(args);
+    if (typeof options.port === 'number' || typeof options.port === 'string') {
+      validatePort(options.port, 'options.port');
+    }
+
+    if (!('port' in options)) {
+      throw new ERR_INVALID_ARG_VALUE(
+        'options',
+        options,
+        'must have the property "port"'
+      );
+    }
+
+    if (this.port != null) {
+      throw new ERR_SERVER_ALREADY_LISTEN();
+    }
+
+    if (portMapper.has(Number(options.port))) {
+      throw new Error(`Port ${options.port} is already in use`);
+    }
+
+    if (callback !== null) {
+      this.once('listening', callback as (...args: unknown[]) => unknown);
+    }
+
+    this.port = Number(options.port);
+    portMapper.set(this.port, { fetch: this.#onRequest.bind(this) });
+    callback?.();
```

Comment:
Does Node.js call this immediately or should this be defered with a `nextTick()`?